From 2796b54a2c6fa92657bd37adf2cac8e52ae71f97 Mon Sep 17 00:00:00 2001
From: Christian Kastner <debian@kvr.at>
Date: Sun, 20 Mar 2011 21:43:06 +0100
Subject: [PATCH] uscan: ZIP repacking fixes

Don't include the current working directory in the tarball, and give ownership
to root instead of leaking the user's UID
---
 scripts/uscan.pl |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/scripts/uscan.pl b/scripts/uscan.pl
index 6b5e4fb..8b146b2 100755
--- a/scripts/uscan.pl
+++ b/scripts/uscan.pl
@@ -1369,8 +1369,14 @@ EOF
 
 	my $newfile_base_gz = "$1.tar.gz";
 	my $tempdir = tempdir ( "uscanXXXX", TMPDIR => 1, CLEANUP => 1 );
-	system("unzip -q -a -d $tempdir $destdir/$newfile_base; GZIP=-9 tar -C $tempdir -czf $destdir/$newfile_base_gz .") == 0 
-	  or die("Repacking from zip to tar.gz failed\n");
+	my $globpattern = "*";
+	system("unzip -q -a -d $tempdir $destdir/$newfile_base") == 0
+	  or die("Repacking from zip to tar.gz failed (could not unzip)\n");
+	if (system("ls -d $tempdir/.[!.]* >/dev/null 2>&1") == 0) {
+	    $globpattern .= " .[!.]*";
+	}
+	system("cd $tempdir; GZIP=-9 tar --owner=root --group=root --mode=a+rX -czf $destdir/$newfile_base_gz $globpattern") == 0
+	  or die("Repacking from zip to tar.gz failed (could not create tarball)\n");
 	unlink "$destdir/$newfile_base";
 	$newfile_base = $newfile_base_gz;
     }
-- 
1.7.4.1

